lib/schema/uuid.js

```diff
@@ -0,0 +1,329 @@
+const Binary = MongooseBuffer.Binary;
+
+/**
+ * Helper function to convert the input hex-string to a buffer
+ * @param {String} hex The hex string to convert
+ * @returns {Buffer} The hex as buffer
+ * @api private
+ */
+
+function hex2buffer(hex) {
+  // use buffer built-in function to convert from hex-string to buffer
+  const buff = Buffer.from(hex, 'hex');
+  return buff;
+}
+
+/**
+ * Helper function to convert the buffer input to a string
+ * @param {Buffer} buf The buffer to convert to a hex-string
+ * @returns {String} The buffer as a hex-string
+ * @api private
+ */
+
+function binary2hex(buf) {
+  // use buffer built-in function to convert from buffer to hex-string
+  const hex = buf.toString('hex');
+  return hex;
+}
+
+/**
+ * Convert a String to Binary
+ * @param {String} uuidStr The value to process
+ * @returns {MongooseBuffer} The binary to store
+ * @api private
+ */
+
+function stringToBinary(uuidStr) {
+  // Protect against undefined & throwing err
+  if (typeof uuidStr !== 'string') uuidStr = '';
+  const hex = uuidStr.replace(/[{}-]/g, ''); // remove extra characters
+  const bytes = hex2buffer(hex);
+  const buff = new MongooseBuffer(bytes);
+  buff._subtype = 4;
+
+  return buff;
+}
+
+/**
+ * Convert binary to a uuid string
+ * @param {Buffer|Binary|String} uuidBin The value to process
+ * @returns {String} The completed uuid-string
+ * @api private
+ */
+function binaryToString(uuidBin) {
+  // i(hasezoey) dont quite know why, but "uuidBin" may sometimes also be the already processed string
+  let hex;
+  if (typeof uuidBin !== 'string') {
+    hex = binary2hex(uuidBin);
+    const uuidStr = hex.substring(0, 8) + '-' + hex.substring(8, 8 + 4) + '-' + hex.substring(12, 12 + 4) + '-' + hex.substring(16, 16 + 4) + '-' + hex.substring(20, 20 + 12);
+    return uuidStr;
+  }
+  return uuidBin;
+}
+
+/**
+ * UUIDv1 SchemaType constructor.
+ *
+ * @param {String} key
+ * @param {Object} options
+ * @inherits SchemaType
+ * @api public
+ */
+
+function SchemaUUID(key, options) {
+  SchemaType.call(this, key, options, 'UUID');
+  this.getters.push(binaryToString);
+}
+
+/**
+ * This schema type's name, to defend against minifiers that mangle
+ * function names.
+ *
+ * @api public
+ */
+SchemaUUID.schemaName = 'UUID';
+
+SchemaUUID.defaultOptions = {};
+
+/*!
+ * Inherits from SchemaType.
+ */
+SchemaUUID.prototype = Object.create(SchemaType.prototype);
+SchemaUUID.prototype.constructor = SchemaUUID;
+
+/*!
+ * ignore
+ */
+
+SchemaUUID._cast = function(value) {
+  if (value === null) {
+    return value;
+  }
+
+  function newBuffer(initbuff) {
+    const buff = new MongooseBuffer(initbuff);
+    buff._subtype = 4;
+    return buff;
+  }
+
+  if (typeof value === 'string') {
+    if (UUID_FORMAT.test(value)) {
+      return stringToBinary(value);
+    } else {
+      throw new CastError(SchemaUUID.schemaName, value, this.path);
+    }
+  }
+
+  if (Buffer.isBuffer(value)) {
```

Comment:
Should we check for the buffer's subtype here? Any reason why we might want to avoid casting buffers and `Binary` instances with different subtypes as UUID?

Reply:
at least at that point it cannot be checked, because those are nodejs native buffers which dont have a bson sub-type, but in the `instanceof Binary` check it can be done, should it be added as another nested check (`if () { if() { throw Error } }`) or directly in the same if (`if (here) {}`)? if nested, what error should be used?